gpAux/gpperfmon/src/gpmon/gpmon_agg.c

```diff
@@ -505,11 +479,8 @@ apr_status_t agg_dup(agg_t** retagg, agg_t* oldagg, apr_pool_t* parent_pool, apr
 		apr_int32_t age = newagg->generation - dp->last_updated_generation - 1;
 		if (age > 0)
 		{
-			if (  (status != GPMON_QLOG_STATUS_SUBMIT
-			       && status != GPMON_QLOG_STATUS_CANCELING
-			       && status != GPMON_QLOG_STATUS_START)
-			   || ((age % 5 == 0) /* don't call is_query_not_active every time because it's expensive */
-			       && is_query_not_active(dp->qlog.key.tmid, dp->qlog.key.ssid, dp->qlog.key.ccnt, active_query_tab, newagg->pool)))
+			if (status == GPMON_QLOG_STATUS_DONE || status == GPMON_QLOG_STATUS_ERROR ||
+			    !is_session_active(dp->qlog.key.ssid, active_session_tab, newagg->pool))
```

Comment:
1. do we really need the is_session_active function? in what situation do we want to skip a request that has not yet completed?
2. in the previous version of the code we also skip: GPMON_QLOG_STATUS_SILENT and GPMON_QLOG_STATUS_INVALID. why don't we need to skip them now?

Reply:
1. I'm not sure if we need it - initially I kept only the part that checks ssid for the case of the sudden session crash; however we figured out later that backend is guaranteed to set the GPMON_QLOG_STATUS_ERROR in that case. Overall, many places in gpperfmon are questionable, but we don't want to rewrite it entirely and try to keep changes to minimum, so I kept the ssid check.
2. Looks like I overlooked these statuses, as they are never set by backend. GPMON_QLOG_STATUS_SILENT is never used at all though.

Reply:
if we want to stick to the old logic, maybe we should try to do a substring search instead of a comparison in `is_query_not_active`?

Reply:
Original query has `string_agg` which won't let the substitution pass

Reply:
Consider the following scenario: a user runs a long query, and for some reason his backend receives SIGKILL. It won't be able to set the final status for gpmon, thus the session is dead, but it's query info files won't be removed and would hang in queries_now.

> however we figured out later that backend is guaranteed to set the GPMON_QLOG_STATUS_ERROR for the case of the sudden session crash
So after reconsidering, this should be read as "for the case of the sudden session crash **catched by PG_TRY/PG_EXCEPT**" - we can correctly set query status in case of query execution error (so basically I misinformed you on this). Nuking the process or (probably) segfaulting may leave the ghost query in the queries_now without ssid checks.

Deleting `is_session_active()` implies deleting `get_active_sessions()`, so the amount of changes would double or triple